feat(oss-iac): adds first cut at OSS IaC article (#59)
masterpointio/masterpoint.io: content/blog/oss-iac.md

And who aspires to that?

There are very, very few Platform or DevOps teams out there with enough staff to fully support the number of application engineers they are expected to. The idea that such teams should use valuable time to write every line of IaC code for their organization is untenable.

In this post, we’ll examine the reasons why you, the platform engineer, should be using OSS IaC and highlight options out there.

## The Why

First, let’s break down why we believe OSS IaC is so critical. Reasons include:

1. Best practices out of the box
2. Security hardened infrastructure

> 3. Less maintenance for fixes + features

Reviewer: I'd rephase to something like:
```suggestion
3. Less maintenance thanks to community fixes and shared experience
```

To use my favorite house building analogy, think about the choice of building a house from scratch vs building it with prefab components. If you’re doing everything from scratch, you have to make sure that every wall, pipe, and wire is built and placed exactly to spec. But when building with prefabricated walls, pipes, and wiring, you only have to connect the various components and the house is move-in ready.

It helps with smaller projects too. Need to renovate that bathroom? That’s a lot simpler when everything already follows a specification. Who knows, there might be upgraded components available that meet your needs.

While LOC is not the silver bullet when assessing software, it can be used as one component of approximating codebase complexity. As an example for this reason to use OSS, let’s look at the number of LOC in some very commonly used modules:

1. [https://github.com/cloudposse/terraform-aws-vpc](https://github.com/cloudposse/terraform-aws-vpc) — 666 LOC

> 2. [https://github.com/cloudposse/terraform-aws-alb—](https://github.com/cloudposse/terraform-aws-alb—) 1064 LOC

Reviewer 2: I confirmed we can to correct the alb url. The new one is validate.

3. [https://github.com/cloudposse/terraform-aws-eks-cluster](https://github.com/cloudposse/terraform-aws-eks-cluster) — 1224 LOC

What’s more efficient? For you to write (and maintain!) those 2K lines of code for your organization or for you to consume those 3 modules and benefit from the hard work that others have put into making them great?

> ## Objections to open-source IaC

Reviewer 2: **I think we need to include an objection related to malicious code.** 
Maybe "Open Source modules could contain malicious code"? Having worked in a few corporate environments, former colleagues would share their distrust of running someone else's code on their boxes and would point at a [Linux PR](https://thehackernews.com/2024/04/malicious-code-in-xz-utils-for-linux.html) or some low level dependency where a bad actor tried to get code injection through a PR.

**Maybe add a section about low quality code?**
I think another hesitation organizations have is a fear that free code is low quality code. Or, you get what you pay for. I think this belief was more prevalent pre 2015? and has largely changed, but I think it's worth highlighting as we interact with mid and large size companies that might lean more corporate. This also transitions well into the next section, "How to evaluate good OSS IaC"

> ## Objections to open-source IaC

Author: ```suggestion
## Objections to open source IaC
```

Of course, no solution is perfect. When I often suggest using OSS modules, I tend to hear these objections and will offer a brief rebuttal to each:

#### “There are too many inputs to so-and-so module – [it feels like a leaky abstraction](https://en.wikipedia.org/wiki/Leaky_abstraction)”

This can happen and there are good examples of OSS modules who are prime offenders. I wouldn’t recommend using them. Evaluate each module independently, review which inputs are required, and make an informed decision. OSS modules aren’t magic pixie dust–you still have to do your due diligence.

#### “Using external modules sounds like dependency hell"

Software is all about managing dependencies, whether OSS or not. The IaC ecosystem is not the Javascript ecosystem; If you use good modules, pin them to a specific version, and update when you need to then you won’t run into the dependency issues that you may have elsewhere.

#### “We can’t use open-source in my organization because of XYZ reason”

> This is a sad reality for some folks and to those folks I suggest: Copy / pasta the open-source module into your codebase, abiding by the license. Plan to update it from time to time.

Author: ```suggestion
This is a sad reality for some folks and to those folks I suggest: Copy / pasta the open source module into your codebase, abiding by the license. Plan to update it from time to time.
```

#### “But I can generate all this IaC with AI… why wouldn’t I do that?”

Did you take the time to understand the AI generated IaC code at a deep level? Did the person who reviewed it? If not, now you have a bunch of code that you need to maintain **and** it won’t be getting updates from the community. Congrats, you’ve increased your future maintenance burden!

> ## How to evaluate good OSS IaC

Reviewer: I’d also add a point about the README content - does it include clear usage instructions, contribution guidelines, examples, contact info for maintainers or developers, etc?

Author: Good call 👍 Will do.

Above I suggest that you need to do your own due diligence to ensure that the modules that you use in your organization are quality. Here are some questions that you can ask to check the quality of any given OSS module:

> #### Check the feature support

Reviewer: Maybe make it a numbered list? This way it's easier to follow the thoughts (IMO).

Author: I originally had it this way and then moved away from it. I think we keep it this way because we're expanding on each point and using headers for each here will allow someone to skim if they don't really care about this section, but want to see how we think about it.

#### Check the number of required `variable` inputs

Some modules have hundreds of inputs, and that can be a sign of poor module design, but is not always. What is more important is the number of inputs that are required. Does the module require you to supply a ton of values for it to work or does it have reasonable defaults?

#### Ensure the module has tests and follows good practices

Good modules use some combination of `terratest`, native TF testing, security checks like `trivy` or `checkov`, `terraform-docs`, and/or a pre-commit workflow. Is the developer of this module using those tools and testing their code on every PR?

#### Investigate if there is a community behind the module

Does the module have a good number of GitHub stars (e.g. 20+)? Are there recent releases and conversation in the PRs or issues? Is there an active Slack or Discord community you can go to if you have questions or concerns? Does the OSS developer or organization have other modules that they support?

Answering these questions can help you understand the quality of the module, which will affect how helpful it will be for you to adopt and use long term.

## Conclusion

> We recommend OSS for IaC because it brings together best practices, strong security, reduced maintenance burden, and a significant drop in infrastructure complexity—benefits that empower teams to work smarter and scale faster. It takes time and effort to learn which OSS modules are right for your organization, but we hope this post gives you the right mindset to evaluate and use those modules.
> 
> Adopting open source IaC doesn't just help you get the job done faster—it empowers platform teams to move with greater autonomy, iterate more freely, and focus their energy on innovation rather than reinvention.

Reviewer: I feel like this can be shortened:
```suggestion
We recommend OSS for IaC because it brings together best practices, strong security, reduced maintenance, and lower complexity — enabling teams to scale faster and focus their energy on innovation rather than reinvention. While it takes effort to choose the right modules, we hope this post helps you approach that process with the right mindset.
```

Author: Love it @gberenice -- Thanks!

> To that end, if you want some great open-source module libraries to check out, start here:

Author: ```suggestion
To that end, if you want some great open source module libraries to check out, start here:
```

---
visible: true
draft: false
title: "Your Infra Isn't Special: Why Open Source Infrastructure as Code (IaC) Wins"
author: Matt Gowie
slug: why-open-source-iac-wins
date: 2025-05-01
description: "If your platform team is still writing every Terraform and OpenTofu resource by hand, you're falling behind. This post explains how open source Infrastructure as Code (IaC) enables modern platform engineering at scale — reducing effort, increasing security, and speeding up delivery."
image: /img/updates/oss-iac/oss-iac.png

> callout: <p>👋 <b>If you're ready to bring OSS best practices into your IaC codebase and workflows, we can help you get there—faster, safer, and with less overhead. <a href='/contact'>Get in touch!</a></p>

Reviewer: ```suggestion
callout: <p>👋 <b>If you're ready to bring OSS best practices into your IaC codebase and workflows, we can help you get there—faster, safer, and with less overhead.</b> <a href='/contact'>Get in touch!</a></p>
```

Of course, no solution is perfect. When I often suggest using OSS modules, I tend to hear these objections and will offer a brief rebuttal to each:

### “There are too many inputs to so-and-so module – [it feels like a leaky abstraction](https://en.wikipedia.org/wiki/Leaky_abstraction)”

This can happen and there are good examples of OSS modules who are prime offenders. I wouldn't recommend using them. Evaluate each module independently, review which inputs are required, and make an informed decision. OSS modules aren't magic pixie dust–you still have to do your due diligence.

### “Using external modules sounds like dependency hell"

Software is all about managing dependencies, whether OSS or not. The IaC ecosystem is not the Javascript ecosystem; If you use good modules, pin them to a specific version, and update when you need to then you won't run into the dependency issues that you may have elsewhere.

### “We can't use open source in my organization because of XYZ reason”

This is a sad reality for some folks and to those folks I suggest: Copy / pasta the open source module into your codebase, abiding by the license. Plan to update it from time to time.

> ### "Open source may have malicious code in it -- what if we get hacked?"

Reviewer: ```suggestion
### "Open source may have malicious code in it – what if we get hacked?"
```

### Less Maintenance

This benefit has two flavors: bug fixes and new features.

When you write all of your IaC code from scratch, it's guaranteed that you're going to need to find and fix any bug that crops up. You'll also need to add features; you might be the platform engineer to enhance the IaC code to add IPv6 support when that day comes (I hear it is just around the corner 😁).

When you're using open source modules, you often find exactly what you're looking for because the community has already:

- run into a particular issue
- thought through it
- and provided a fix or turnkey way to implement it

This community effort means the fix for your issue is only a version bump away.

IPv6 is a _GREAT_ example. Are you ready for all of your public endpoints to be IPv6? I know that not all of our projects and client work is. Of course we're not the only ones!

> But when IPv6 support is required, I know that [great contributions like this one](https://github.com/cloudposse/terraform-aws-alb/pull/186) will make it much easier to implement IPv6 support. We'll be able to lean on the efforts of the community instead of having to do days of in-the-trenches development work to make IPv6 happen.

Reviewer 3: ```suggestion
But when IPv6 support is required, I know that [great contributions like this one](https://github.com/cloudposse/terraform-aws-alb/pull/186) will make it much easier to implement IPv6 support. We'll be able to lean on the efforts of the community instead of having to do days of in-the-trenches development work to make IPv6 happen. This directly translates to lower operational costs and frees up an organization's most expensive resource—engineers' time—to focus on building product features that drive value, not on reinventing a solved problem.
```

Adding a quick sentence on further emphasizing on why it matters. Might be redundant, see if it fits well?

### Check the feature support

Does it default to support your needs or can you configure it to support your needs? Does it meet 95% of your needs and a quick fork + commit will get it to 100%?

### What version of TF does it support

Is your version of TF supported and can you use the module without upgrading? Does it make sense to upgrade if you need to for this use-case?

### Evaluate the scope of the module

Is it too big and includes dozens of resources that you don't understand, won't use, and are for use-cases that you're not interested in? Is it too small and simply wraps a single resource without any supporting logic that provides value? Or, goldilocks style, is the size just right?

### Check the number of required `variable` inputs

Some modules have hundreds of inputs, and that can be a sign of poor module design, but is not always. What is more important is the number of inputs that are required. Does the module require you to supply a ton of values for it to work or does it have reasonable defaults?

### Ensure the module has tests and follows good practices

> Good modules use some combination of `terratest`, native TF testing, security checks like `trivy` or `checkov`, `terraform-docs`, and/or a pre-commit workflow. Is the developer of this module using those tools and testing their code on every PR?

Reviewer 3: ```suggestion
Good modules use some combination of `terratest`, [native TF testing](https://developer.hashicorp.com/terraform/language/tests), security checks like `trivy` or `checkov`, `terraform-docs`, and/or a pre-commit workflow. Is the developer of this module using those tools and testing their code on every PR?
```

Author: Using our own link instead of Hashis! Success! 

```suggestion
Good modules use some combination of `terratest`, [native TF testing](https://masterpoint.io/blog/terraform-opentofu-terminology-breakdown/#native-tf-testing), security checks like `trivy` or `checkov`, `terraform-docs`, and/or a pre-commit workflow. Is the developer of this module using those tools and testing their code on every PR?
```
